### ISSUE TITLE
Doesn't work with Tensorflow 2 and Python 3.8

### ISSUE THREAD

**Issue by georgekolasa** [MTG/essentia#1167]
Is there any way to get the Essentia [music genre classification (link)](https://mtg.github.io/essentia-labs/news/tensorflow/2019/10/19/tensorflow-models-in-essentia/) to work with Python3.8  and Tensorflow 2? I've tried everything. 

It seems like you guys are still using Tensorflow 1.15 and there's not really a way to use the most current version of Python (since Tensorflow 1.15 doesn't support Python 3.8). I need Python 3.8 for the project I'm working on and can't really just downgrade for just this one thing. Any way around this?

**Comment by palonso** [comment]
the easiest way to do this is to link against libtensorflow instead of the Python wheel. This can be achieved with the `setup_from_libtensorflow.sh` script. First, modify it to use the TensorFlow version you need:
https://github.com/MTG/essentia/blob/e3b6cfeaed4ea92cd7377be2a0980aa91b08b378/src/3rdparty/tensorflow/setup_from_libtensorflow.sh#L11
Then run the script and build Essentia following steps 5 to 8 from the blog post you mentioned.

**Comment by georgekolasa** [comment]
@palonso  Thanks for your response.

When I try to run step 6 ( python3 waf configure --build-static --with-python --with-tensorflow ) in the folder with my waf file (manually installed from source), I get this error:

waf: error: no such option: --build-static

Why is this happening?

**Comment by palonso** [comment]
` python3 waf configure --build-static --with-python --with-tensorflow` works well in my system.

Are you running it from the base of the Essentia folder?
When you say your waf file, is it something else than the one that ships with Essentia?

**Comment by georgekolasa** [comment]
Okay, so that worked, thank you @palonso 

But now it's saying "name 'TensorflowPredict' is not defined" when I try  to run the model from the guide. I've had this similar error before and TensorflowPredict seems to be part of essentia. Is there any reason why this wouldn't show? I'm using Tensorflow 2.5.0 and didn't have any major errors on following those install steps discussed above.

**Comment by palonso** [comment]
Good to hear that it worked!

Your new error sounds like you are using Essentia without TensorFlow. Maybe the interpreter is loading a previously installed version from other source, such as pip?
If this is the case uninstall it with `pip uninstall essentia`

With the steps you followed, Essentia should be installed in `/usr/local/lib/python3/dist-packages/`. 
Can you confirm it? 

You can see which installation is being imported  by doing:
```python
import essentia
essentia.__file__
```

Finally, can you check if  `/usr/local/lib/python3/dist-packages/` is in the interpreter's path?
```python
import sys
sys.path 
```

**Comment by georgekolasa** [comment]
@palonso  Thanks! So that worked. The only thing is it takes a ton of manual work to set up. We're trying to use essentia and it's models as part of our music visualization project (github [link here](https://github.com/cbaddeley/Synesthesia)). We want to use things like the danceability and mood models to determine how our image is drawn. 

Is there an easier way to do this that works with Python3.8 and Tensorflow 2, like a pip package we can include in the install requires part of our setup.py? pip install essentia and pip install essentia-tensorflow seem to have issues when either installing or running the model (See here : https://gyazo.com/f1ee6e5578f10dc8ed32cefecd0c030d )

**Comment by dbogdanov** [comment]
The issue with the wheels for TensorFlow 2 is that we are getting a clash when running `import tensorflow` and `import essentia.standard` together, and [it is not resolved yet](https://github.com/tensorflow/tensorflow/issues/). We are pending on the TensorFlow team to find a workaround. 

For prototyping, you can try the [test Essentia wheels with TensorFlow 2 we did](https://drive.google.com/drive/folders/1Ipl48pcZAvdX-ogYk0cx2gwwkDE58rbX), but you will get into this issue if you need to import both libraries.

**Comment by seth814** [comment]
I don't think this addresses the core issue of running tensorflow 2 alongside python 3.8 AND essentia.
I was in the process of updating my Dockerfile.
This will let you run essentia-tensorflow with python 3.8 from a base tensorflow 2.8 image.
I uninstall tensorflow here since I only care about essentia-tensorflow that is on pypi.
Seems to be accessing the GPU and letting me run all my neural nets through essentia.

Dockerfile
```
FROM tensorflow/tensorflow:2.8.0-gpu

# essentia
RUN apt update -y
RUN rm -rf /var/lib/apt/lists/*
RUN apt clean -y
RUN apt update -y
RUN apt upgrade -y
RUN apt install -y --no-install-recommends autoconf automake pkg-config libtool
RUN pip3 install --no-cache -U pip
RUN pip install --no-cache pip-autoremove
RUN pip-autoremove tensorflow -y
RUN pip3 install --no-cache essentia-tensorflow
```

Left out a lot of the other stuff from my Dockerfile, but this is the top of it.
`RUN apt install -y --no-install-recommends autoconf automake pkg-config libtool`
This is probably not needed. Needed it for other libraries I was installing.

**Comment by palonso** [comment]
The new essentia-tensorflow wheels are shipped with a monolithic build of TensorFlow 2.5 that doesn't interfere with TensorFlow's python package, which means that essentia-tensorflow is now compatible with any TensorFlow version in Python 3.7 to 3.9.
As the TensorFlow library is embedded in the wheels, it is not required to have it pre-installed in the system in any way.

Note that the user needs to install CUDA 11.2 and cuDNN 8.1 ([required](https://www.tensorflow.org/install/source#gpu) by TF 2.5) to use essentia-tensorflow with a GPU. To use essentia-tensorflow and TensorFlow with a GPU simultaneously, it is recommended to use TensorFLow >= 2.5  to have a single CUDA/cuDNN dependency.
